testng-core/src/test/java/test/dataprovider/DataProviderTest.java

```diff
@@ -666,6 +671,48 @@ public Object[][] getSuiteFileNames() {
     };
   }
 
+  @Test(description = "GITHUB-3045")
+  public void testIfDataProviderListenerInvokedOnlyOncePerDataProvider() {
+    runTest(DataProviderTestClassSample.class, "");
+  }
+
+  @Test(description = "GITHUB-3045")
+  public void testIfDataProviderListenerInvokedOnlyOncePerDataProviderWhenListenerAddedViaSuite() {
+    runTest(DataProviderWithoutListenerTestClassSample.class, DataProviderListener.class.getName());
+  }
+
+  private static void runTest(Class<?> clazz, String listener) {
```

Comment:
Class<? extends TestngListener> could be expressive.

Reply:
No. The first parameter is actually the test class and not the listener. The listener is being passed in as a String

Reply:
Yes. That's my point. Why do you use a string?

Reply:
TestNG is either ways going to accept a string. So why not ?

Reply:
It is a question of style and api design. You should not exposed untyped object when it is possible to avoid it.
String is not a strong type because you can put everything into it.
Here Class is a better choice and you will not have to check inputs.
Just do the string conversion the later you can.

Reply:
@juherr - Couple of things

* This is an existing API in the `XmlSuite` class and I am not adding a new one. `XmlSuite` will only accept strings.
* The method in question is being used by 2 test methods with the sole intention of avoiding code duplication. So the rationale of exposing any untyped object or api design does not hold valid here.
* I will still have to check for inputs, because the reusable method `runTest` is going to set a fully qualified class name to `XmlSuite` ONLY if a listener was passed. So if a `Class<?>` was passed, then I would be sending in null and thus doing a null check and if a string is being sent, i do a empty check.

So I am not convinced that the change you are suggesting is going to add any value here because this is just within the scope of a reusable method that is being shared by 2 test methods with the sole intention of avoiding code duplication.

Reply:
@juherr - It wasn't an argument. I was genuinely trying to understand the rationale. I agree with you on the fact that it would need to be a general good habit. My contention was that in this case it wasn't needed.

Either ways, it turns out that we both missed one part in that method :) I was using a string to pass the listener name, but was actually resorting to using a hard coded listener name always. I have now fixed it by replacing it with a boolean which indicates if a listener should or should not be wired in or not.

Reply:
My apologies if it sounded as if I was on an argumentative mode. That was not my intention.